hera_cal/smooth_cal.py

```diff
@@ -67,6 +68,84 @@ def single_iterative_fft_dly(gains, wgts, freqs, conv_crit=1e-5, maxiter=100):
+    to filter calibration solutions along the time and frequency axes simultaneously.
+
+    Arguments:
+        freqs: ndarray of frequency channels in Hz
+        times: ndarray of shape=(Ntimes) of Julian dates as floats in units of days
+        freq_scale: frequency scale in MHz to use for the low-pass filter. freq_scale^-1 corresponds
+            to the half-width (i.e. the width of the positive part) of the region in fourier
+            space, symmetric about 0, that is retained after filtering.
+            Note that freq_scale is in MHz while freqs is in Hz.
+        time_scale: time scale in seconds. Defined analogously to freq_scale.
+            Note that time_scale is in seconds, times is in days.
+        eigenval_cutoff: sinc_matrix eigenvalue cutoff to use for included dpss modes.
+            Only used when the filtering method is 'DPSS'
+        return_1D_filters: if True, return the 1D filters that make up the 2D DPSS filter.
+            Default is False
+
+    Returns:
+        filters: DPSS filtering vectors, ndarray of size (Ntimes * Nfreqs, N_frequency_vectors * N_time_vectors)
+    """
+    delay_scale = (freq_scale * 1e6) ** -1  # Puts it in seconds
+    fringe_scale = (time_scale) ** -1  # fringe scale in Hz
+    time_in_seconds = (times - times.min()) * 60 * 60 * 24  # time array in seconds
+
+    freq_filters, nfreq_comps = uvtools.dspec.dpss_operator(
+        freqs,
+        filter_centers=[0],
+        filter_half_widths=[delay_scale],
+        eigenval_cutoff=[eigenval_cutoff],
+    )  # DPSS filters along the frequency-axis
+    freq_filters = freq_filters.real.astype(np.float32)
+
+    time_filters, ntime_comps = uvtools.dspec.dpss_operator(
+        time_in_seconds,
+        filter_centers=[0],
+        filter_half_widths=[fringe_scale],
+        eigenval_cutoff=[eigenval_cutoff],
+    )  # DPSS filters along the time-axis
+    time_filters = time_filters.real.astype(np.float32)
+
+    filters = np.reshape(
+        time_filters[:, None, :, None] * freq_filters[None, :, None, :],
+        (times.shape[0] * freqs.shape[0], ntime_comps[0] * nfreq_comps[0]),
+    )  # 2D time-frequency DPSS vectors
+    if return_1D_filters:
+        return filters, freq_filters, time_filters
+
+    else:
+        return filters
+
+
+def fit_solution_matrix(weights, design_matrix):
+    """
+    Calculate the linear least squares solution matrix
+    from a design matrix, A and a weights matrix W
+    S = [A^T W A]^{-1} A^T W
+
+    Arguments:
+        weights: ndarray of shape (Ntimes, Nfreqs) of calibration flags
+        design_matrix: ndarray of shape (Ntimes * Nfreqs, N_frequency_vectors * N_time_vectors)
+            obtained from hera_cal.smooth_cal.dpss_filters
+
+    Returns:
+        S: ndarray of shape (N_frequency_vectors * N_time_vectors, )
+    """
+    cmat = (design_matrix.T * weights.reshape(-1)) @ design_matrix
+
+    if np.linalg.cond(cmat) >= 1e9:
```

Comment:
No magic numbers. This should be a default kwarg.

Reply:
Also, computing the condition number can be expensive. Is doing this check significantly increasing runtime? Why not just always do pinv?

Reply:
> No magic numbers. This should be a default kwarg.

"magic numbers" makes me think of this because there are numbers and wizards

<img width="410" alt="image" src="https://user-images.githubusercontent.com/6164284/160531456-29b275ad-548e-4731-9499-8848725f669c.png">

Reply:
I updated this to just perform the pinv instead of checking the condition number. This function is a slightly modified version of a function in `uvtools` which is where the `1e9` came from. We might want to make some updates there as well